### DISCUSSION TITLE
how to enable pickle for Ensemble objects

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#207] [Design & Development]
In our last call I discussed the fact that I found the Ensemble objects do not currently support pickle.  We have to have the capability because some algorithms already pass Ensemble objects so this is a clear and glaring hole.   You mentioned in the call that you thought that because we had pickle defined for all the pieces that go into an ensemble we could just pickle the individual pieces and reassemble them.  I am not so sure about that as I facing two very different dilemmas:

1. I can find no example in the pybind11 documents nor in an online search that shows if it is possible to run pickle inside the pybind11 C++ wrapper code.   The section in the documentation on running python in a C++ program hints that is possible, but the way they describe that capability is very different from this.  I'm working that trying that approach could be a dark alley that will be hard to get out of without injury.  If you know something I do not, however, please educate me.  Web searches are always a bit of a crap shoot for this kind of thing.
2. Even if we could use pickle, the standard approach in pybind11 concerns me for a potential memory problem.  That is, what we do for existing objects is serialize the pieces of an object like Seismogram that is constructed from multiple base classes and put the pieces in a tuple.  The same tuple is taken apart in the reverse operation.  For Seismogram and TimeSeries that would rarely if every be a memory issue but it could easily create an issue for Ensembles.   Case in point are the miniseed data event files I download that are all lower 48 broadbands with 4000 s long records.  Each file is about 700 Mby.  Expand that by about a factor of 8 and you have a lot of data to handle in memory.  Modern computers have huge memories, but many could not handle two copies of a 5 GB file in memory.  It might get even worse with pickle as it could actually expand the size a fair amount.   

We need to discuss some solutions to this before I enter one of those dark alleys.   Perhaps them more fundamental problem is how one controls you dask and spark handle serialization and what control we have over what they do?  Large object serialization must be something they support.   In fact, a quick web search found [this page](https://distributed.dask.org/en/latest/serialization.html) on dask alternatives.   Before I dive too deep in this I think we need a brief discussion of the topic.

**Comment by wangyinz** [comment]
Aha, I have run into this before, so I know the answer: yes, you can use pickle within a C++ code to serialize any python objects. You can actually find the usage in our old serialization method of Metadata:
https://github.com/wangyinz/mspass/blob/cc10c4d7d720c67cb9d250d8dbbf27648487e78b/cxx/src/lib/utility/Metadata.cc#L261-L274
Of course, this part of the code is already abandoned, but you can do something similar for Ensemble and serialize all its members.

The memory problem is actually a completely different problem. I guess it is always a concern to serialize any objects, and if the object is large enough, you can always run out of memory no matter how efficient the serialization is implemented. This actually brings us back to the classic question: what is the granularity of parallelism. I could be that a 5 GB ensemble is just not the right atomic data for our system.  I guess that's also part of the reason why we did not implement serialization for Ensembles.

Anyway, no matter how we think the granularity should be, we would always want to optimize the serialization part of MsPASS, and I think the link you provide is among the things we should explore. I guess it is not essential until we hit the limit with some real workflow.

**Reply by pavlis** [reply]
That was very helpful.   I had forgotten you learned that trick.

Agreed we need to put the memory issue on the back burner for now.   Yet another example of an axiom of software development:  "Make it work before you make it fast(efficient)." 

A couple followup questions I have thought that you may or may not be able to answer.  If you don't know right off from the previous experience tell me so and I'll do the web hunting.   

First, I think no matter what I do I have to create a tuple in one of two forms:
```
( serialized_ensemble_metadata, serialized_error_log, ensemble_is_live_boolean, serialized_data (all together as one huge block)
```
or
```
(serialized_ensemble_metadata, serialized_error_log, ensemble_is_live_boolean, n, serialized_member_0, ... , serialized_member[n-1])
```
I'm not sure which of the above is even feasible - see followup questions below:

1.  I'm a little confused about what we depricated.   I think the proper use is to use serialize_metadata_py and restore_serialized_metadata_py (defined in Metadata.cc) for the ensemble Metadata.   I guess I could use pickle, but I think that would just add another layer?  The pybind11 code uses those two functions for metadata anyway - that's why I say using pickle would just add a layer right?
2. I'm going to have to use boost serialziation for the ErrorLogger component of the ensemble anyway.   I has no pickle defined because with TimeSeries and Seismogram elog is part of ProcessingHistory - recall we had problems and stuck with that design a few months ago. 
3. Now the real sticky point about serializing the data.  I think the key question is a subtlety of the distinction between a tuple and an array.   All the pybind11 examples pass the data around as a tuple, but I wonder if that is required?   Do you know if the pickle wrapper lines in pybind11 require a tuple to exchange the data?   They all use this kind of contruct for the deserializating lambda:  `[](py::tuple t) ` but is that required?  I.e. could I equally use `[](py::list t)`?   Why that matter is that a tuple is immutable and code to serialize the data array in model 2 (serialize each member individually) has an issue.   I, at least, don't know how to create a tuple from a list without first building the list.  i.e. the only way I know how to create a tuple in the pybind11 code to pass out of pickle is to first create a list , call it *data_list*, and end the C++ pybind11 lambda code with `return(make_tuple(data_list))`.  Maybe that just does a const cast on *data_list* and doesn't make another copy, but I have no idea how to find out one way or the other.   Model 1, by the way, (serializing all the data into one big buffer) always has the triple copy problem.   i.e. the binding code has to set aside monster buffer and serialize all the members into that buffer.   I fear that is all very confusing, but part of the reason is I'm confused.   If you don't get it all the absolutely key question is this:  is a py::tuple required as the data structure for serialization and deserialization with pickle as implemented in pybind11?

This alley is getting darker as I get into it.

**Comment by wangyinz** [comment]
See answers below:

1. Well, my bad. I got it confused. We are actually using pickle in serialize_metadata_py.
https://github.com/wangyinz/mspass/blob/cc10c4d7d720c67cb9d250d8dbbf27648487e78b/cxx/src/lib/utility/Metadata.cc#L288-L300
Anyway, yes, we probably should just call the serialize_metadata_py and restore_serialized_metadata_py directly.

2. Yes, we should use boost serialization here. That is exactly how we implemented it for ProcessingHistory: https://github.com/wangyinz/mspass/blob/0a61bf1fbbcfea0de30c15738cba4a974095577b/cxx/python/utility/utility_py.cc#L890-L904

3. I think a tuple is perfectly fine here. We just need to build a tuple whose last element is a list:
```( serialized_ensemble_metadata, serialized_error_log, ensemble_is_live_boolean, member_list)```
The `member_list` can be supported by pickle if we call it like this: `pickle.dumps(list(Ensemble.member))`. Basically, we can just convert the `TimeSeriesVector` or `SeismogramVector` into a list directly. I don't know how pybind11 got it implemented, but I would assume it is efficient unless we prove otherwise. In that case, we will need to implement an efficient serialization for vectors, where we could just dump the whole data region I guess.

**Reply by pavlis** [reply]
Ok, thanks.  That was very helpful.  Now I have to see if I can build something that (a) compiles, and (b) the works.   I presume tests for the pickle interface should go in the testpickle.py
